File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKAction.java
package br.com.insanitech.spritekit;

import android.content.Context;
import android.support.annotation.IdRes;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;
import java.util.Random;

import br.com.insanitech.spritekit.logger.Logger;

public abstract class SKAction implements Cloneable {
    private static SKActionTimingMode defaultTiming = SKActionTimingMode.Linear;

    public static void setDefaultTiming(SKActionTimingMode mode) {
        defaultTiming = mode;
    }

    WeakReference<SKNode> parent = null;

    String key = "";
    Runnable completion;
    boolean started = false;
    long startedTime = 0;

    public float speed = 0;
    public long duration = 1000;
    public SKActionTimingMode timingMode = defaultTiming;

    public SKAction reverseAction() {
        return null;
    }

    void setParent(SKNode node) {
        if (node != null) {
            parent = new WeakReference<SKNode>(node);
        } else {
            parent = null;
        }
    }

    SKNode getParent() {
        return parent.get();
    }

    void start() {
        if (!started) {
            started = true;
            startedTime = System.currentTimeMillis();
            computeStart();
        }
    }

    abstract void computeStart();

    void computeAction() {
        long elapsed = System.currentTimeMillis() - startedTime;
        if (parent != null && parent.get() != null) {
            computeAction(elapsed);
        }
        checkCompleted(elapsed);
    }

    abstract void computeAction(long elapsed);

    abstract void computeFinish();

    abstract boolean willHandleFinish();

    boolean checkCompleted(long elapsed) {
        if (!willHandleFinish() && elapsed >= duration) {
            if (parent != null) {
                computeFinish();
                parent.get().actionCompleted(this);
            }
            dispatchCompletion();
            parent = null;
            return true;
        }
        return false;
    }

    void dispatchCompletion() {
        if (completion != null) {
            completion.run();
        }
    }

    public SKAction copy() throws CloneNotSupportedException {
        SKAction copy = (SKAction) clone();
        copy.duration = duration;
        copy.key = key;
        copy.speed = speed;
        copy.started = started;
        copy.startedTime = startedTime;
        copy.timingMode = timingMode;
        copy.completion = completion;
        return copy;
    }

    public static SKAction moveBy(SKPoint deltaPosition, long duration) {
        return moveBy(deltaPosition.x, deltaPosition.y, duration);
    }

    public static SKAction moveBy(float deltaX, float deltaY, long duration) {
        SKAction action = new SKActionMoveBy(new SKPoint(deltaX, deltaY));
        action.duration = duration;
        return action;
    }

    public static SKAction moveTo(float toX, float toY, long duration) {
        return moveTo(new SKPoint(toX, toY), duration);
    }

    public static SKAction moveTo(SKPoint position, long duration) {
        SKAction action = new SKActionMoveTo(position);
        action.duration = duration;
        return action;
    }

    public static SKAction rotateByAngle(float radians, long duration) {
        SKAction action = new SKActionRotateBy(radians);
        action.duration = duration;
        return action;
    }

    public static SKAction rotateToAngle(float radians, long duration) {
        SKAction action = new SKActionRotateTo(radians);
        action.duration = duration;
        return action;
    }

    public static SKAction resizeBy(float width, float height, long duration) {
        return resizeBy(new SKSize(width, height), duration);
    }

    public static SKAction resizeBy(SKSize size, long duration) {
        SKAction action = new SKActionResizeBy(size);
        action.duration = duration;
        return action;
    }

    public static SKAction resizeTo(float width, float height, long duration) {
        return resizeTo(new SKSize(width, height), duration);
    }

    public static SKAction resizeTo(SKSize size, long duration) {
        SKAction action = new SKActionResizeTo(size);
        action.duration = duration;
        return action;
    }

    public static SKAction playSoundFile(Context context, @IdRes int resId) {
        final SKAction action = new SKActionPlaySoundFile(context, resId);
        action.duration = 0;
        return action;
    }

    public static SKAction sequence(List<SKAction> sequence) {
        return new SKActionSequence(new LinkedList<SKAction>(sequence));
    }

    public static SKAction group(List<SKAction> group) {
        return new SKActionGroup(new ArrayList<SKAction>(group));
    }

    public static SKAction waitFor(long duration) {
        SKAction action = new SKActionWaitFor();
        action.duration = duration;
        return action;
    }

    public static SKAction waitFor(long duration, long range) {
        SKAction action = new SKActionWaitFor();
        Random random = new Random();
        long sum = (long) random.nextInt((int) range);
        action.duration = duration + sum;
        return action;
    }

    public static SKAction fadeIn(long duration) {
        SKAction action = new SKActionFadeIn();
        action.duration = duration;
        return action;
    }

    public static SKAction fadeOut(long duration) {
        SKAction action = new SKActionFadeOut();
        action.duration = duration;
        return action;
    }

    public static SKAction fadeAlphaTo(float alpha, long duration) {
        SKAction action = new SKActionFadeAlphaTo(alpha);
        action.duration = duration;
        return action;
    }

    public static SKAction fadeAlphaBy(float alpha, long duration) {
        SKAction action = new SKActionFadeAlphaBy(alpha);
        action.duration = duration;
        return action;
    }

    public static SKAction scaleTo(float scale, long duration) {
        return scaleTo(scale, scale, duration);
    }

    public static SKAction scaleTo(float x, float y, long duration) {
        SKAction action = new SKActionScaleTo(x, y);
        action.duration = duration;
        return action;
    }

    public static SKAction scaleBy(float scale, long duration) {
        return scaleBy(scale, scale, duration);
    }

    public static SKAction scaleBy(float x, float y, long duration) {
        SKAction action = new SKActionScaleBy(x, y);
        action.duration = duration;
        return action;
    }

    public static SKAction setTexture(SKTexture texture) {
        SKAction action = new SKActionSetTexture(texture);
        action.duration = 0;
        return action;
    }

    public static SKAction run(Runnable completion) {
        SKAction action = new SKActionRun(completion);
        action.duration = 0;
        return action;
    }

    public static SKAction removeFromParent() {
        SKAction action = new SKActionRemoveFromParent();
        action.duration = 0;
        return action;
    }

    void log(String content, Object... args) {
        Logger.log(getClass().getSimpleName(), String.format(Locale.getDefault(), content, args));
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKAction.kt
package br.com.insanitech.spritekit

import android.content.Context
import android.support.annotation.IdRes

import java.lang.ref.WeakReference
import java.util.ArrayList
import java.util.LinkedList
import java.util.Locale
import java.util.Random

import br.com.insanitech.spritekit.logger.Logger

abstract class SKAction : Cloneable {
    private var started = false
    private var startedTime: Long = 0

    internal var key = ""
    internal var parent: SKNode? = null
    internal var completion: SKBlock? = null

    var speed = 1f
    var duration: Long = 1000
    var timingMode = defaultTiming

    fun reverseAction(): SKAction? = null

    internal abstract fun computeStart()

    internal fun computeAction() {
        if (!started) {
            started = true
            startedTime = System.currentTimeMillis()
            computeStart()
        }

        val elapsed = System.currentTimeMillis() - startedTime
        computeAction(elapsed)
        checkAndCompleteIfFinished(elapsed)
    }

    internal abstract fun computeAction(elapsed: Long)

    internal abstract fun computeFinish()

    internal fun checkAndCompleteIfFinished(elapsed: Long) {
        if (elapsed >= this.duration) {
            this.computeFinish()
            this.removeFromParent()
            this.dispatchCompletion()
        }
    }

    internal fun removeFromParent() {
        this.parent?.actions?.remove(this)
        this.parent = null
    }

    internal fun dispatchCompletion() {
        this.completion?.invoke()
    }

    @Throws(CloneNotSupportedException::class)
    fun copy(): SKAction {
        val copy = clone() as SKAction
        copy.duration = duration
        copy.key = key
        copy.speed = speed
        copy.started = started
        copy.startedTime = startedTime
        copy.timingMode = timingMode
        copy.completion = completion
        return copy
    }

    companion object {
        private var defaultTiming = SKActionTimingMode.Linear

        fun moveBy(deltaPosition: SKPoint, duration: Long): SKAction =
                moveBy(deltaPosition.x, deltaPosition.y, duration)

        fun moveBy(deltaX: Float, deltaY: Float, duration: Long): SKAction {
            val action = SKActionMoveBy(SKPoint(deltaX, deltaY))
            action.duration = duration
            return action
        }

        fun moveTo(toX: Float, toY: Float, duration: Long): SKAction =
                moveTo(SKPoint(toX, toY), duration)

        fun moveTo(position: SKPoint, duration: Long): SKAction {
            val action = SKActionMoveTo(position)
            action.duration = duration
            return action
        }

        fun rotateByAngle(radians: Float, duration: Long): SKAction {
            val action = SKActionRotateBy(radians)
            action.duration = duration
            return action
        }

        fun rotateToAngle(radians: Float, duration: Long): SKAction {
            val action = SKActionRotateTo(radians)
            action.duration = duration
            return action
        }

        fun resizeBy(width: Float, height: Float, duration: Long): SKAction =
                resizeBy(SKSize(width, height), duration)

        fun resizeBy(size: SKSize, duration: Long): SKAction {
            val action = SKActionResizeBy(size)
            action.duration = duration
            return action
        }

        fun resizeTo(width: Float, height: Float, duration: Long): SKAction =
                resizeTo(SKSize(width, height), duration)

        fun resizeTo(size: SKSize, duration: Long): SKAction {
            val action = SKActionResizeTo(size)
            action.duration = duration
            return action
        }

        fun playSoundFile(context: Context, @IdRes resId: Int): SKAction {
            val action = SKActionPlaySoundFile(context, resId)
            action.duration = 0
            return action
        }

        fun sequence(sequence: List<SKAction>): SKAction = SKActionSequence(LinkedList(sequence))

        fun group(group: List<SKAction>): SKAction = SKActionGroup(ArrayList(group))

        fun waitFor(duration: Long): SKAction {
            val action = SKActionWaitFor()
            action.duration = duration
            return action
        }

        fun waitFor(duration: Long, range: Long): SKAction {
            val action = SKActionWaitFor()
            val random = Random()
            val sum = random.nextInt(range.toInt()).toLong()
            action.duration = duration + sum
            return action
        }

        fun fadeIn(duration: Long): SKAction {
            val action = SKActionFadeIn()
            action.duration = duration
            return action
        }

        fun fadeOut(duration: Long): SKAction {
            val action = SKActionFadeOut()
            action.duration = duration
            return action
        }

        fun fadeAlphaTo(alpha: Float, duration: Long): SKAction {
            val action = SKActionFadeAlphaTo(alpha)
            action.duration = duration
            return action
        }

        fun fadeAlphaBy(alpha: Float, duration: Long): SKAction {
            val action = SKActionFadeAlphaBy(alpha)
            action.duration = duration
            return action
        }

        fun scaleTo(scale: Float, duration: Long): SKAction = scaleTo(scale, scale, duration)

        fun scaleTo(x: Float, y: Float, duration: Long): SKAction {
            val action = SKActionScaleTo(x, y)
            action.duration = duration
            return action
        }

        fun scaleBy(scale: Float, duration: Long): SKAction = scaleBy(scale, scale, duration)

        fun scaleBy(x: Float, y: Float, duration: Long): SKAction {
            val action = SKActionScaleBy(x, y)
            action.duration = duration
            return action
        }

        fun setTexture(texture: SKTexture): SKAction {
            val action = SKActionSetTexture(texture)
            action.duration = 0
            return action
        }

        fun run(completion: SKBlock): SKAction {
            val action = SKActionRun(completion)
            action.duration = 0
            return action
        }

        fun removeFromParent(): SKAction {
            val action = SKActionRemoveFromParent()
            action.duration = 0
            return action
        }
    }
}